### ISSUE TITLE
Using load_custom overrides all other holiday definitions

### ISSUE THREAD

**Issue by nello** [holidays/holidays#251]
Hi,

Firstly apologies if I'm somehow using the library incorrectly, but I'm seeing very weird behaviour when I try to add in some extra holidays by using load_custom().

If I use Holidays.between(Date.today, Date.today + 1.years, :au_vic) to return the holidays for the next year for Victoria, Australia I get a nice list of holidays:

```
irb(main):002:0> Holidays.between(Date.today, Date.today + 1.years, :au_vic)
=> [{:date=>Sun, 25 Dec 2016, :name=>"Christmas Day", :regions=>[:au, :au_qld, :au_nsw, :au_act, :au_tas, :au_wa]}, {:date=>Mon, 26 Dec 2016, :name=>"Boxing Day", :regions=>[:au, :au_nsw, :au_vic, :au_qld, :au_act, :au_wa]}, {:date=>Tue, 27 Dec 2016, :name=>"Christmas Day", :regions=>[:au_vic, :au_nt]}, {:date=>Sun, 01 Jan 2017, :name=>"New Year's Day", :regions=>[:au]}, {:date=>Thu, 26 Jan 2017, :name=>"Australia Day", :regions=>[:au]}, {:date=>Mon, 13 Mar 2017, :name=>"Labour Day", :regions=>[:au_vic]}, {:date=>Fri, 14 Apr 2017, :name=>"Good Friday", :regions=>[:au]}, {:date=>Sat, 15 Apr 2017, :name=>"Easter Saturday", :regions=>[:au_nsw, :au_vic, :au_qld, :au_nt, :au_act, :au_sa]}, {:date=>Sun, 16 Apr 2017, :name=>"Easter Sunday", :regions=>[:au_nsw, :au_vic]}, {:date=>Mon, 17 Apr 2017, :name=>"Easter Monday", :regions=>[:au]}, {:date=>Tue, 25 Apr 2017, :name=>"ANZAC Day", :regions=>[:au]}, {:date=>Tue, 25 Apr 2017, :name=>"ANZAC Day", :regions=>[:au_nsw, :au_vic, :au_qld, :au_nt, :au_act, :au_sa, :au_tas]}, {:date=>Mon, 12 Jun 2017, :name=>"Queen's Birthday", :regions=>[:au_act, :au_nsw, :au_sa, :au_tas, :au_nt, :au_vic]}]
```

That unfortunately is slightly wrong because the definition for the Friday before the AFL Grand Final (don't ask!) is incorrect. Strangely the test for that holiday has a different year to the one in the actual rule, but as I actually need the 2017 holiday anyway, that's not a big deal.

After reading the docs, I created a local yaml definition file for that holiday which parses and loads successfully:

```
months:
  10:
  - name: Fridays before the AFL Grand Final
    regions: [au_vic]
    function: afl_grand_finals(year)


methods:
  afl_grand_finals:
    arguments: year
    source: |
      case year
      when 2015
        Date.civil(2015, 10, 2)
      when 2016
        Date.civil(2016, 9, 30)
      when 2017
        Date.civil(2017, 9, 29)
      end
```

But if I try to use this definition:
```
irb(main):002:0> Holidays.load_custom('config/holidays.yml')
=> {10=>[{:name=>"Fridays before the AFL Grand Final", :regions=>[:au_vic], :function=>"afl_grand_finals(year)", :function_arguments=>[:year]}]}

irb(main):002:0> Holidays.between(Date.today, Date.today + 1.years, :au_vic)
=> [{:date=>Fri, 29 Sep 2017, :name=>"Fridays before the AFL Grand Final", :regions=>[:au_vic]}]
```
my new holiday is the **only one** returned. This doesn't seem to be expected behaviour!

Any clues would be much appreciated. I have tried naming my definition both "afl_grand_final" (which is a clash for an existing rule in the au definitions) and "afl_grand_finals" but neither seems to work.

Thanks for your help!

**Comment by nello** [comment]
Ah! I have a clue. If load_custom() is the first call to the library, it doesn't seem to load the default rules. I think that's just a bug, right?

**Comment by ptrimble** [comment]
Hi @nello! Thanks for your patience. I haven't had a chance to look at this but I hope to do it soon. My wife had a baby in December and I only just now getting back onto some kind of schedule that will allow dev work. I'll get back to you!

**Comment by williamweckl** [comment]
+1 Having the same problem here.

**Comment by MDes41** [comment]
This issue seems to be closed @williamweckl, I am trying to recreate and I am getting the correct date I assume.  

```2.3.0 :032 >  Holidays.between(Date.today, Date.today.next_year, :au_vic)
 => [{:date=>#<Date: 2017-03-13 ((2457826j,0s,0n),+0s,2299161j)>, :name=>"Labour Day", :regions=>[:au_vic]}, {:date=>#<Date: 2017-04-14 ((2457858j,0s,0n),+0s,2299161j)>, :name=>"Good Friday", :regions=>[:au]}, {:date=>#<Date: 2017-04-15 ((2457859j,0s,0n),+0s,2299161j)>, :name=>"Easter Saturday", :regions=>[:au_nsw, :au_vic, :au_qld, :au_nt, :au_act, :au_sa]}, {:date=>#<Date: 2017-04-16 ((2457860j,0s,0n),+0s,2299161j)>, :name=>"Easter Sunday", :regions=>[:au_nsw, :au_vic]}, {:date=>#<Date: 2017-04-17 ((2457861j,0s,0n),+0s,2299161j)>, :name=>"Easter Monday", :regions=>[:au]}, {:date=>#<Date: 2017-04-25 ((2457869j,0s,0n),+0s,2299161j)>, :name=>"ANZAC Day", :regions=>[:au_nsw, :au_vic, :au_qld, :au_nt, :au_act, :au_sa, :au_tas]}, {:date=>#<Date: 2017-04-25 ((2457869j,0s,0n),+0s,2299161j)>, :name=>"ANZAC Day", :regions=>[:au]}, {:date=>#<Date: 2017-06-12 ((2457917j,0s,0n),+0s,2299161j)>, :name=>"Queen's Birthday", :regions=>[:au_act, :au_nsw, :au_sa, :au_tas, :au_nt, :au_vic]}, {:date=>#<Date: 2017-09-29 ((2458026j,0s,0n),+0s,2299161j)>, :name=>"Friday before the AFL Grand Final", :regions=>[:au_vic]}, {:date=>#<Date: 2017-12-25 ((2458113j,0s,0n),+0s,2299161j)>, :name=>"Christmas Day", :regions=>[:au_qld, :au_nsw, :au_act, :au_tas, :au_wa, :au_vic, :au_nt]}, {:date=>#<Date: 2017-12-26 ((2458114j,0s,0n),+0s,2299161j)>, :name=>"Boxing Day", :regions=>[:au_nsw, :au_vic, :au_qld, :au_act, :au_wa]}, {:date=>#<Date: 2018-01-01 ((2458120j,0s,0n),+0s,2299161j)>, :name=>"New Year's Day", :regions=>[:au]}, {:date=>#<Date: 2018-01-26 ((2458145j,0s,0n),+0s,2299161j)>, :name=>"Australia Day", :regions=>[:au]}]```


```2.3.0 :033 >  Holidays.between(Date.today, Date.today.next_year, :au_vic).find { |day| day[:name] == "Friday before the AFL Grand Final" }
 => {:date=>#<Date: 2017-09-29 ((2458026j,0s,0n),+0s,2299161j)>, :name=>"Friday before the AFL Grand Final", :regions=>[:au_vic]}```

Are you still having the issue?

**Comment by MDes41** [comment]
I recreated the issue and fixed it by adding ```Factory::Finder.between.call(Date.today, Date.today, :au)``` to top of the method in load custom.  I assume you could make a blank file and have the same functionality as long as you add a dummy region.  Seems like a hack but seems to fix the problem

**Comment by ppeble** [comment]
Okay, I see the issue. I am going to write it down here so I won't forget it later in case I can't fix it tonight.

The relevant bit of code is here: https://github.com/holidays/holidays/blob/master/lib/holidays/finder/context/parse_options.rb#L53-L67

Unfortunately `parse_options` is one of the handful sections of essentially unrefactored code that still exists in the app. This and the code generation are mostly the original code, save a few bits. That means that this issue probably existed from the moment `load_custom` was introduced. For what it is worth, **I** was the one that wrote `load_custom` before I took over this repo so I probably introduced this way back when. 😄 

When we load the custom file, everything is loaded into memory. The `:au_vic` region is loaded. Everyone is happy.

Then we make the `between` call:

```ruby
Holidays.between(Date.today, Date.civil(2018, 3, 18), :au_vic)
```

(I'm not in rails/active_support so I just hardcoded the date)

When deciding whether to 'load' a region, we check whether it already exists: https://github.com/holidays/holidays/blob/master/lib/holidays/finder/context/parse_options.rb#L54

In the case of our loaded custom region, the answer is a resounding YES. It exists! No need to try to load anything else. Which is...bad. We want to load everything in `:au` as well! That's what is missing.

If we were to run this BEFORE we loaded anything custom, the `load_definition_data` method would throw a `LoadError` because there is no definition file for 'au_vic'. Upon a `LoadError` we would run this: https://github.com/holidays/holidays/blob/master/lib/holidays/finder/context/parse_options.rb#L60-L64

This would basically say "Okay, this is 'au_vic', now get the overall region, which is au, and search for that". Then we would load the rest of the holidays and get what we expect.

So, in short: **we need to check whether we are in sub-region regardless of whether the region is already loaded**. This is the only way to get reliable results.

I am going to see if I can address this tonight. Hit me up if you have questions or if you spot a mistake!

**Comment by ppeble** [comment]
I haven't forgotten about this! In fact I think I'll have a PR shortly. I have the problem solved and at this point I am simply trying to break my new implementation.

**Comment by ppeble** [comment]
Hi everyone!

I am confident that https://github.com/holidays/holidays/pull/271 fixes this issue. I will be rolling out v5.5.0 shortly, please give that a try. If you notice that I have _not_ fixed the issue please post it here and I will reopen.

Note: I tested locally using the example provided and things seem good. So let's see! Thanks for your patience!

**Comment by robwold** [comment]
@ppeble I've run into this issue more or less exactly as far as I can tell, could there have been a regression here?
I'm trying to set it up in a rail's app's `config/initializers/holidays.rb` and have had to use the same workaround discussed above:
```ruby
Holidays.between(Date.today, Date.today + 1.years, :gb) # without this, standard holidays are being overwritten by the next line
Holidays.load_custom(File.join(File.dirname(__FILE__), 'custom_holidays.yaml'))
```
where
`config/initializers/holidays/custom_holidays.yaml` looks like
```yaml
months:
  9:
  - name: "Queen's funeral"
    regions: [gb]
    mday: 19
    year_ranges:
      limited: [2022]
```
